### PR TITLE
Add generic Docker workflow

### PR DESCRIPTION
**Pull request by typerat**

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+          credentials_json: ${{ secrets.gar_service_account_key }}
+
+      - name: Login to GAR
+        run: |
+          docker login -u _json_key \
+          --password-stdin ${{ inputs.registry }} < ${{ steps.auth.outputs.credentials_file_path }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
```
Comment (hollow) [inline .github/workflows/docker.yml:77]: what's the reason for hardcoding the commit ref of the metadata-action here?

Comment (typerat) [reply .github/workflows/docker.yml:77]: @tundeaoni said it should be this way for security reasons

Comment (typerat) [reply .github/workflows/docker.yml]: We do need to upgrade it though, as it's using deprecated features

**Comment by typerat** [comment]
TODO: change input defaults (after creating the final resources in GAR)

File: .github/workflows/docker.yml (hunk 1 above)
Comment (typerat) [reply .github/workflows/docker.yml]: @tundeaoni FYI

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    runs-on: ubuntu-latest
+    outputs:
+      image: ${{ steps.meta.outputs.tags }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup SSH
+        uses: webfactory/ssh-agent@v0.5.0
+        env:
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
```
Comment (hollow) [inline .github/workflows/docker.yml:60]: why is this env var neede here?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      - name: Authenticate to Google Cloud
+        id: auth
+        uses: google-github-actions/auth@v0
+        with:
+          credentials_json: ${{ secrets.gar_service_account_key }}
+
+      - name: Login to GAR
+        run: |
+          docker login -u _json_key \
+          --password-stdin ${{ inputs.registry }} < ${{ steps.auth.outputs.credentials_file_path }}
```
Comment (hollow) [inline .github/workflows/docker.yml:76]: Have you checked the [docker login action](https://github.com/docker/login-action#google-artifact-registry-gar)? 

This also reminds me of [SRE-187](https://remerge.atlassian.net/browse/SRE-187), but I don't want us to get distracted now 🤔

Comment (typerat) [reply .github/workflows/docker.yml]: Oh, I've even used it in the past, just forgot it existed. However, I just noticed that the password is printed if we use that.
image

I'd leave this as it is and revisit it in a new branch for SRE-187. What do you think?

File: .github/workflows/docker.yml (hunk 2 above)
Comment (typerat) [reply .github/workflows/docker.yml]: Thanks for noticing, it isn't needed -> removed 👍

**Review by tundeaoni — APPROVED** [review]

**Pull request merged**